keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2169 @@
+
+    // DrainPlan is executed from the first entry to the last entry during the Drain stage.
+    // DrainPlanEntry podType fields should be in the following order:
+    // nil -> DaemonSet -> Static
+    // DrainPlanEntry priority fields should be in ascending order for each podType.
+    // If the priority and podType are the same, concrete selectors are executed first.
+    //
+    // The following entries are injected into the drainPlan on the NodeMaintenance admission:
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "Default"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "Default"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "Default"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "Default"
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "DaemonSet"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "DaemonSet"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "DaemonSet"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "DaemonSet"
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "Static"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "Static"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "Static"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "Static"
+    //
+    // Duplicate entries are not allowed.
+    // This field is immutable.
+    DrainPlan []DrainPlanEntry
+
+    // Reason for the maintenance.
+    Reason string
+}
+
+const (
+// Default selects all pods except DaemonSet and Static pods.	
+Default PodType = "Default"
+// DaemonSet selects DaemonSet pods.
+DaemonSet PodType = "DaemonSet"
+// Static selects static pods.
+Static PodType = "Static"
+)
+
+type DrainPlanEntry struct {
+    // PodSelector selects pods according to their labels.
+    // This can help to select which pods of the same priority should be evacuated first.
+    // +optional
+    PodSelector *metav1.LabelSelector
+    // PodPriority specifies a pod priority.
+    // Pods with a priority less or equal to this value are selected.
+    PodPriority int32
+    // PodType selects pods according to the pod type:
+    // - Default selects all pods except DaemonSet and Static pods.	
+    // - DaemonSet selects DaemonSet pods.
+    // - Static selects static pods.
+    PodType PodType
+}
+
+type NodeMaintenanceStatus struct {
+    // StageStatuses tracks the statuses of started stages.
+    StageStatuses []StageStatus
+    // List of a maintenance statuses for all nodes targeted by this maintenance.
+    NodeStatuses []NodeMaintenanceNodeStatus
```

Comment:
@thockin I do not like the pattern of using status as a database

how do we offer consistency , is there only one actor managing this? 
If the status is wiped out by mistake are we able to rebuild the current state?

Reply:
is this going to support 5000 nodes per example?

Reply:
>  If the status is wiped out by mistake are we able to rebuild the current state?

- Kubernetes no longer expects `.status` to be rebuildable
- I think it is reconstructable anyway; start from scratch (as if the kubelet had not previously been running)

Reply:
> is this going to support 5000 nodes per example?

Good question.
We could add fields into Node status if we want the kubelet to report something per-node (and just report it, not using `.status` to record cluster administrator intent.
We could also add a SomethingSlice API like we did with Endpoints, but that does not feel like a good fit.

Reply:
+1, this is problematic:

I am trying to bend the status somehow, but it is not feasible.

We cannot even limit how many nodes the selectors should select on admission, since new nodes can appear during the NodeMaintenance lifetime (especially Idle). 

We could introduce a MultiNodeMaintenance object, whose only point would be to create x number of NodeMaintenance objects, one for each node. And it would synchronize the current stage between itself and the children.

But yeah, storing this in the Node status probably sounds like the best solution.

Reply:
This should be solved in the current API design.